### PR TITLE
LineCountBear.py: Add min_lines_per_file

### PR DESCRIPTION
**Pull request by damian1996**
Adds the lower limit for length of file, which is set to 0 by default.

Closes https://github.com/coala/coala-bears/issues/1477

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: bears/general/LineCountBear.py [hunk 1]
```diff
@@ -18,15 +19,35 @@ def _get_blank_line_count(self, file):
...
-        smaller than a given size.
+        Count the number of lines in a file and ensure that they lie within
+        the range of given sizes. There are files for which lower limit isn't
+        required, but these files still shouldn't be rejected. Some of such
+        files when minimum is 1 are mentioned in ignore_files_regex
+        in the repository .coafile.
 
-        :param max_lines_per_file: Number of lines allowed per file.
-        :param exclude_blank_lines: "True" if blank lines are to be excluded.
+        :param max_lines_per_file: Maximum number of lines allowed per file.
```
Comment (newbazz) [inline bears/general/LineCountBear.py]: Indent the docstring properly ;P

File: tests/general/LineCountBearTest.py [hunk 2]
```diff
@@ -40,3 +42,57 @@ def test_exclude_blank_lines(self):
...
+            self.uut, [''],
+            [Result.from_values('LineCountBear',
+                                'This file had 0 lines, which is 1 lines '
+                                'fewer than the minimum required.',
+                                severity=RESULT_SEVERITY.NORMAL,
+                                file='default')],
+            filename='default')
+        self.check_results(self.uut, [''], [], filename='__init__.py')
+
+    def test_max_limit_is_smaller_than_min_limit(self):
```
Comment (newbazz) [inline tests/general/LineCountBearTest.py]: Write a test case in which you test any of the ignored files ;)

Comment (damian1996) [reply tests/general/LineCountBearTest.py]: I added such test case :) 

`self.check_results(self.uut, [''], [], filename='__init__.py')`

This test case should check that __init__.py file is empty and despite that min_lines_per_file=1 and max_lines_per_file=2 this test case shouldn't write any message.

Comment (newbazz) [reply tests/general/LineCountBearTest.py]: but there is coverage error in the CI.

File: .coafile [hunk 3]
```diff
@@ -48,6 +48,7 @@ body_enforce_issue_reference = False
 enabled = False
 bears = LineCountBear
 max_lines_per_file = 1000
+ignore_files_regex = ^.*__init__.py$, ^.*\.gitkeep$, ^.*\.gitignore$
```
Comment (newbazz) [inline .coafile]: Are you sure you need to ignore init files, coz IMO they have some code written in them.

Comment (newbazz) [reply .coafile]: Yeah ik but not all init files are empty you know ;)

File: bears/general/LineCountBear.py [hunk 4]
```diff
@@ -39,3 +64,14 @@ def run(self, filename, file, max_lines_per_file: int,
                                  extra=file_length-max_lines_per_file)),
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+        elif file_length < min_lines_per_file:
+            if not any(regex.search(filename) for regex in ignore_regexes):
+                yield Result.from_values(
+                    origin=self,
+                    message=('This file had {count} lines, which is {missing} '
+                             'lines fewer than the minimum required.'
+                             .format(count=file_length,
```
Comment (newbazz) [inline bears/general/LineCountBear.py]: No need of using `count` and `missing`.

File: .coafile (hunk 3 above)
Comment (damian1996) [reply .coafile]: @newbazz But I don't ignore all init.py files :) I will ignore it just in special case, it means when number of lines in init file is smaller than allowed minimum set. Then ignore file, means that for this type of file, I can't write the right message, because init can have less lines than set minimum.

File: bears/general/LineCountBear.py (hunk 4 above)
Comment (newbazz) [reply bears/general/LineCountBear.py]: change this as well.

File: bears/general/LineCountBear.py [hunk 5]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
         return num_blank_lines
 
     def run(self, filename, file, max_lines_per_file: int,
-            exclude_blank_lines: bool=False):
+            ignore_files_regex: typed_list(str),
+            exclude_blank_lines: bool=False,
```
Comment (newbazz) [inline bears/general/LineCountBear.py]: Change its default value to True ;P

Comment (newbazz) [reply bears/general/LineCountBear.py]: btw blank lines as in trailing blank lines?

Comment (damian1996) [reply bears/general/LineCountBear.py]: @newbazz Exclude_blank_lines was written by someone else. This is part of code which already was written here when I started doing this issue. :) So I wouldn't want change this :)

Comment (damian1996) [reply bears/general/LineCountBear.py]: @newbazz And blank lines are allow by default, so this should be set to True :)

File: bears/general/LineCountBear.py [hunk 6]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
         return num_blank_lines
 
     def run(self, filename, file, max_lines_per_file: int,
-            exclude_blank_lines: bool=False):
+            ignore_files_regex: typed_list(str),
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: there's `ignore_files` defined by coala's settings system. You shouldn't implement such settings for bears.

File: bears/general/LineCountBear.py [hunk 7]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
+            exclude_blank_lines: bool=False,
+            min_lines_per_file: int=1):
         """
-        Count the number of lines in a file and ensure that they are
-        smaller than a given size.
+        Count the number of lines in a file and ensure that they lie within
+        the range of given sizes. There are files for which lower limit isn't
+        required, but these files still shouldn't be rejected. Some of such
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: if a limit is not required, then the files won't be rejected all times, right? This is kinda confusing...

File: bears/general/LineCountBear.py [hunk 8]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
+        files when minimum is 1 are mentioned in ignore_files_regex
+        in the repository .coafile.
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: This is specific to this repo, please don't document such infos in a bear.

File: bears/general/LineCountBear.py [hunk 9]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
-        :param max_lines_per_file: Number of lines allowed per file.
-        :param exclude_blank_lines: "True" if blank lines are to be excluded.
+        :param max_lines_per_file:
+            Maximum number of lines allowed per file.
+        :param exclude_blank_lines:
+            ``True`` if blank lines are to be excluded.
+        :param min_lines_per_file:
+            Minimum number of lines required per file.
+        :param ignore_files_regex:
+            Files, for which minimum is not required.
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Please add another commit changing the docstring layout before for `max_lines_per_file` and `exclude_blank_lines`

File: bears/general/LineCountBear.py [hunk 10]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
+
+        if min_lines_per_file > max_lines_per_file:
+            yield Result.from_values(
+                origin=self,
+                message=('Allowed maximum lines per file({}) is smaller than '
+                         'minimum lines per file({}) and this is prohibited'
+                         .format(max_lines_per_file,
+                                 min_lines_per_file)),
+                severity=RESULT_SEVERITY.NORMAL,
+                file=filename)
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: this is not a `Result`, but an error. Please raise an exception.

File: bears/general/LineCountBear.py [hunk 11]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
+            return
         if exclude_blank_lines:
             num_blank_lines = self._get_blank_line_count(file)
             file_length = file_length - num_blank_lines
         if file_length > max_lines_per_file:
             yield Result.from_values(
                 origin=self,
-                message=('This file had {count} lines, which is {extra} '
+                message=('This file had {} lines, which is {} '
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: why this change?

File: bears/general/LineCountBear.py [hunk 12]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
                          'lines more than the maximum limit specified.'
-                         .format(count=file_length,
-                                 extra=file_length-max_lines_per_file)),
+                         .format(file_length,
+                                 file_length-max_lines_per_file)),
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Definitely unnecessary change. Please file an issue to discuss removing named placeholders if the format-string is simple enough.

File: bears/general/LineCountBear.py [hunk 13]
```diff
@@ -18,24 +19,58 @@ def _get_blank_line_count(self, file):
...
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+        elif file_length < min_lines_per_file:
+            if not any(regex.search(filename) for regex in ignore_regexes):
+                yield Result.from_values(
+                    origin=self,
+                    message=('This file had {} lines, which is {} '
+                             'lines fewer than the minimum required.'
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: than ~~the minimum~~ required.

File: bears/general/LineCountBear.py (hunk 12 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 This have been suggested by previous reviewer :p

File: bears/general/LineCountBear.py (hunk 11 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 This have been suggested by previous reviewer :p

**Comment by Makman2** [comment]
unack bd7812f5496150401b468ed59d1a6981a578c588

File: bears/general/LineCountBear.py (hunk 6 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 Hm, so I shouldn't consider exceptions for required limit? :) I will need ignore_files for check exceptions later :)

File: bears/general/LineCountBear.py (hunk 7 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 There are files which are below lower limit, but the files are still needed in project. For example init.py files.
Maybe I should set min_lines_per_file to 0 and I won't mention any exceptions? Then user can define his own exceptions, if user will have bigger minimum limit than 0 ? :)

File: bears/general/LineCountBear.py (hunk 13 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: Ok, I will correct this

File: bears/general/LineCountBear.py (hunk 12 above)
Comment (Makman2) [reply bears/general/LineCountBear.py]: Not part of your desired change :+1: Could be a good newcomer issue anyway :+1:

File: bears/general/LineCountBear.py (hunk 7 above)
Comment (Makman2) [reply bears/general/LineCountBear.py]: `min_lines_per_file` should be disabled by default (--> `=0`), but I rather meant to improve the phrasing here :)

**Comment by damian1996** [comment]
I removed ignore_files, because I set min_lines_per_file to 0 by default. Additionally, I have had problem with raise ValueError (with test it). I tried to do similar to DartLintBear, where contributors used assertRaisesRegex, however I can't use this by other error, so I made it another (like in the newest commit).

File: bears/general/LineCountBear.py [hunk 14]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
...
+            min_lines_per_file: int=0):
         """
-        Count the number of lines in a file and ensure that they are
-        smaller than a given size.
+        Count the number of lines in a file and ensure that they lie within
+        the range of given sizes.
 
-        :param max_lines_per_file: Number of lines allowed per file.
-        :param exclude_blank_lines: "True" if blank lines are to be excluded.
+        :param max_lines_per_file: Maximum number of lines allowed per file.
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Unnecessary change. You can put this into a separate commit

File: bears/general/LineCountBear.py [hunk 15]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
...
+        :param exclude_blank_lines: ``True`` if blank lines are to be excluded.
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Here too, separate commit

File: .coafile (hunk 3 above)
Comment (Makman2) [reply .coafile]: I don't understand this change too, is the `LineCountBear` triggered in our repo?

Comment (Makman2) [reply .coafile]: Also, we don't have an `ignore_files_regex` setting :O

File: bears/general/LineCountBear.py [hunk 16]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
...
+            raise ValueError('Allowed maximum lines per file({}) is smaller '
+                             'than minimum lines per file({}) and this is '
+                             'prohibited'.format(max_lines_per_file,
+                                                 min_lines_per_file))
+        return
+
     def run(self, filename, file, max_lines_per_file: int,
-            exclude_blank_lines: bool=False):
+            exclude_blank_lines: bool=False,
+            min_lines_per_file: int=0):
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: hm, I would argue that the param `min_lines_per_file` should go before `max_lines_per_file` ;)

File: bears/general/LineCountBear.py [hunk 17]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
             list(filter(lambda x: re.match(r'^\s*$', x), file)))
         return num_blank_lines
 
+    def _check_correct_range(self, min_lines_per_file, max_lines_per_file):
+        if min_lines_per_file > max_lines_per_file:
+            print(min_lines_per_file, max_lines_per_file)
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: debug remain, please remove that

File: bears/general/LineCountBear.py [hunk 18]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
...
+            raise ValueError('Allowed maximum lines per file({}) is smaller '
+                             'than minimum lines per file({}) and this is '
+                             'prohibited'.format(max_lines_per_file,
+                                                 min_lines_per_file))
+        return
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Obsolete return

Comment (Makman2) [inline bears/general/LineCountBear.py]: Tbh, I wouldn't write a special function for that this is a simple "if condition -> raise"

File: bears/general/LineCountBear.py [hunk 19]
```diff
@@ -17,16 +17,37 @@ def _get_blank_line_count(self, file):
...
         file_length = len(file)
+        try:
+            self._check_correct_range(min_lines_per_file, max_lines_per_file)
+        except ValueError as e:
+            yield Result.from_values(
+                origin=self,
+                message=str(e),
+                severity=RESULT_SEVERITY.NORMAL,
+                file=filename)
+            return
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: No that's not a result, let the exception bubble up.

(We maybe have to implement somewhen a special exception for such errors, so the user gets a nice message instead of a traceback... Though probably those checks will then happen inside `__init__` then... Not sure yet)

File: bears/general/LineCountBear.py [hunk 20]
```diff
@@ -39,3 +60,12 @@ def run(self, filename, file, max_lines_per_file: int,
                                  extra=file_length-max_lines_per_file)),
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: please add another blank line, I believe this improves readability

File: bears/general/LineCountBear.py [hunk 21]
```diff
@@ -39,3 +60,12 @@ def run(self, filename, file, max_lines_per_file: int,
...
+        if file_length < min_lines_per_file:
+            yield Result.from_values(
+                origin=self,
+                message=('This file had {count} lines, which is {missing} '
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: It *has* :)

File: bears/general/LineCountBear.py [hunk 22]
```diff
@@ -39,3 +60,12 @@ def run(self, filename, file, max_lines_per_file: int,
...
+                         'lines fewer than required.'
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: You don't have to name the placeholders, just use `{}` twice

File: bears/general/LineCountBear.py [hunk 23]
```diff
@@ -39,3 +60,12 @@ def run(self, filename, file, max_lines_per_file: int,
...
+                         .format(count=file_length,
+                                 missing=min_lines_per_file - file_length)),
+                severity=RESULT_SEVERITY.NORMAL,
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Isn't this default?

File: bears/general/LineCountBear.py [hunk 24]
```diff
@@ -39,3 +60,12 @@ def run(self, filename, file, max_lines_per_file: int,
                                  extra=file_length-max_lines_per_file)),
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+        if file_length < min_lines_per_file:
+            yield Result.from_values(
+                origin=self,
+                message=('This file had {count} lines, which is {missing} '
+                         'lines fewer than required.'
+                         .format(count=file_length,
+                                 missing=min_lines_per_file - file_length)),
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: I would include in the message the given min-lines-per-file setting, so the user remembers what he has passed. Consider you have a few files violating this min-lines-per-file, It's a bit confusing if both numbers change in every result (because all flagged files have a different number of lines)

File: tests/general/LineCountBearTest.py [hunk 25]
```diff
@@ -9,12 +9,17 @@
 
 class LineCountBearTest(LocalBearTestHelper):
 
+    EXCEPTION_MESSAGE_RE = ('ValueError: Allowed maximum lines per file(1) '
+                            'is smaller than minimum lines per file(2) and '
+                            'this is prohibited')
```
Comment (Makman2) [inline tests/general/LineCountBearTest.py]: Unused variable

File: tests/general/LineCountBearTest.py [hunk 26]
```diff
@@ -40,3 +45,53 @@ def test_exclude_blank_lines(self):
...
             filename='default')
+
+    def test_of_range_allowed_for_lines_per_file(self):
+        self.section.append(Setting('max_lines_per_file', 10))
+        self.section.append(Setting('exclude_blank_lines', True))
+        self.section.append(Setting('min_lines_per_file', 5))
+        self.check_results(
+            self.uut, ['line 1', '', 'line 2', '', 'line 3', 'line 4'],
+            [Result.from_values('LineCountBear',
+                                'This file had 4 lines, which is 1 lines '
```
Comment (Makman2) [inline tests/general/LineCountBearTest.py]: hmm, this is a good demonstration why the message needs polishing ;)
1 line*s*

**Comment by Makman2** [comment]
unack 5201237eef64dace3109c080415be21250986006

File: tests/general/LineCountBearTest.py (hunk 26 above)
Comment (damian1996) [reply tests/general/LineCountBearTest.py]: @Makman2 Should I to correct only my tests (for min_lines_per_file)? Maybe should I also to correct tests with max_lines_per_file?

File: bears/general/LineCountBear.py (hunk 19 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 I have a small problem. I made something like this : 

`
if min_lines_per_file > max_lines_per_file:  
 raise ValueError('Allowed maximum lines per file({}) is smaller than minimum lines per file({}) and this is prohibited'.format(max_lines_per_file, min_lines_per_file)) 
`

And I made similar test to DartLintBear with use of assertRaisesRegex. However I had error.
How can I make test for this? :)

Test look like this : 

`def test_max_limit_is_smaller_than_min_limit(self):`

        EXCEPTION_MESSAGE = ('ValueError: Allowed maximum lines per '
                                'file(1) is smaller than minimum lines per '
                                'file(2) and this is prohibited')
        self.section.append(Setting('max_lines_per_file', 1))
        self.section.append(Setting('min_lines_per_file', 2))
        self.section.append(Setting('exclude_blank_lines', True))
        with self.assertRaisesRegex(AssertionError, EXCEPTION_MESSAGE):
            self.check_validity(self.uut, [])`

Comment (Makman2) [reply bears/general/LineCountBear.py]: Tbh I just read a different bear already using `logging.err` + `return`. I've thought about this again, and in fact the logging approach works too, even with caching. So take what you like, I'm gonna think about a uniform way to handle this later...

File: tests/general/LineCountBearTest.py (hunk 26 above)
Comment (Makman2) [reply tests/general/LineCountBearTest.py]: Ah it's the same? Then keep it, we can file an issue afterwards to correct both messages and use a grammatical-proof message :)

File: tests/general/LineCountBearTest.py [hunk 27]
```diff
@@ -1,20 +1,22 @@
...
 from bears.general.LineCountBear import LineCountBear
 from coalib.testing.LocalBearTestHelper import LocalBearTestHelper
+from coalib.testing.LocalBearTestHelper import execute_bear
 from coalib.results.Result import RESULT_SEVERITY, Result
 from coalib.settings.Section import Section
 from coalib.settings.Setting import Setting
 
 
 class LineCountBearTest(LocalBearTestHelper):
-
```
Comment (jayvdb) [inline tests/general/LineCountBearTest.py]: please restore this blank line

File: tests/general/LineCountBearTest.py [hunk 28]
```diff
@@ -1,20 +1,22 @@
...
     def setUp(self):
         self.section = Section('name')
         self.uut = LineCountBear(self.section, Queue())
 
     def test_run(self):
         self.section.append(Setting('max_lines_per_file', 1))
+        self.section.append(Setting('min_lines_per_file', 0))
```
Comment (jayvdb) [inline tests/general/LineCountBearTest.py]: why is this necessary?  it has a default

File: bears/general/LineCountBear.py [hunk 29]
```diff
@@ -17,19 +18,29 @@ def _get_blank_line_count(self, file):
...
             exclude_blank_lines: bool=False):
         """
-        Count the number of lines in a file and ensure that they are
-        smaller than a given size.
+        Count the number of lines in a file and ensure that they lie within
+        the range of given sizes.
 
         :param max_lines_per_file: Number of lines allowed per file.
         :param exclude_blank_lines: "True" if blank lines are to be excluded.
+        :param min_lines_per_file: Minimum number of lines required per file.
```
Comment (jayvdb) [inline bears/general/LineCountBear.py]: keep the order of parameters the same as in the function declaration

**Comment by damian1996** [comment]
@jayvdb I made corrections in my commit :)

File: bears/general/LineCountBear.py [hunk 30]
```diff
@@ -17,19 +18,29 @@ def _get_blank_line_count(self, file):
             list(filter(lambda x: re.match(r'^\s*$', x), file)))
         return num_blank_lines
 
-    def run(self, filename, file, max_lines_per_file: int,
+    def run(self, filename, file, min_lines_per_file: int=0,
+            max_lines_per_file: int=1000,
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: adding a default here is a separate change :+1:

File: bears/general/LineCountBear.py [hunk 31]
```diff
@@ -17,19 +18,29 @@ def _get_blank_line_count(self, file):
...
+            logging.error('Allowed maximum lines per file({}) is smaller than'
+                          ' minimum lines per file({}) and this is '
+                          'prohibited.'.format(max_lines_per_file,
+                                               min_lines_per_file))
+            return
+
         if exclude_blank_lines:
             num_blank_lines = self._get_blank_line_count(file)
             file_length = file_length - num_blank_lines
+
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: separate commit

File: bears/general/LineCountBear.py [hunk 32]
```diff
@@ -39,3 +50,12 @@ def run(self, filename, file, max_lines_per_file: int,
                                  extra=file_length-max_lines_per_file)),
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+
+        if file_length < min_lines_per_file:
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: you can actually make this an `elif` :+1:

**Comment by Makman2** [comment]
unack 7dccef61282c5b88ac744c3fdfe342c6230124cb

File: bears/general/LineCountBear.py [hunk 33]
```diff
@@ -17,16 +18,25 @@ def _get_blank_line_count(self, file):
...
+        the range of given sizes.
 
+        :param min_lines_per_file: Minimum number of lines required per file.
         :param max_lines_per_file: Number of lines allowed per file.
         :param exclude_blank_lines: "True" if blank lines are to be excluded.
         """
         file_length = len(file)
+        if min_lines_per_file > max_lines_per_file:
+            logging.error('Allowed maximum lines per file({}) is smaller than'
+                          ' minimum lines per file({})'
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: add a space between `file` and the braces, looks like some code term otherwise.

File: bears/general/LineCountBear.py [hunk 34]
```diff
@@ -17,16 +18,25 @@ def _get_blank_line_count(self, file):
             list(filter(lambda x: re.match(r'^\s*$', x), file)))
         return num_blank_lines
 
-    def run(self, filename, file, max_lines_per_file: int,
+    def run(self, filename, file, min_lines_per_file: int,
+            max_lines_per_file: int=1000,
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: Still, separate commit for this one

File: bears/general/LineCountBear.py [hunk 35]
```diff
@@ -39,3 +49,12 @@ def run(self, filename, file, max_lines_per_file: int,
                                  extra=file_length-max_lines_per_file)),
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+
+        elif file_length < min_lines_per_file:
+            yield Result.from_values(
+                origin=self,
+                message=('This file has {} lines, which is {} lines fewer '
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: fewer -> less

File: bears/general/LineCountBear.py [hunk 36]
```diff
@@ -39,3 +49,12 @@ def run(self, filename, file, max_lines_per_file: int,
...
+                         'than required.'
+                         .format(file_length,
+                                 min_lines_per_file - file_length)),
```
Comment (Makman2) [inline bears/general/LineCountBear.py]: maybe instead of saying "this is 3 lines less than required", say "x lines are required". That makes more results from the same kind better to read, as this number won't change.

File: bears/general/LineCountBear.py (hunk 34 above)
Comment (damian1996) [reply bears/general/LineCountBear.py]: @Makman2 What do you understand by new commit? :) My second commit in this PR or new issue for newcomers ? :)

File: bears/general/LineCountBear.py [hunk 37]
```diff
@@ -39,3 +50,12 @@ def run(self, filename, file, max_lines_per_file: int,
...
                 severity=RESULT_SEVERITY.NORMAL,
                 file=filename)
+
+        elif file_length < min_lines_per_file:
+            yield Result.from_values(
+                origin=self,
+                message=('This file has {} lines, while {} lines are '
+                         'required.'
+                         .format(file_length,
+                                 min_lines_per_file - file_length)),
```
Comment (HizkiFW) [inline bears/general/LineCountBear.py]: I think the message could be rephrased to:

> This file has `file_length` lines, while `min_lines_per_file` lines are required.

The phrase "n lines are required" specifies the minimum number of lines required, not the number of lines to be added to meet the requirement.

Comment (damian1996) [reply bears/general/LineCountBear.py]: @HizkiFW Thanks, I forgot about change this :) Corrected :)

File: tests/general/LineCountBearTest.py [hunk 38]
```diff
@@ -40,3 +42,52 @@ def test_exclude_blank_lines(self):
                                 severity=RESULT_SEVERITY.NORMAL,
                                 file='default')],
             filename='default')
+
+    def test_of_range_allowed_for_lines_per_file(self):
```
Comment (palash25) [inline tests/general/LineCountBearTest.py]: IMO you should drop the `of` from the function name and rename it to `test_range_allowed_for_lines_per_file`
`test_of` didn't sound like a test function name while I was reading it.

File: tests/general/LineCountBearTest.py [hunk 39]
```diff
@@ -40,3 +42,52 @@ def test_exclude_blank_lines(self):
...
+        self.check_results(
+            self.uut, ['line 1', '', 'line 2', '', 'line 3', 'line 4'],
+            [Result.from_values('LineCountBear',
+                                'This file has 4 lines, while 5 lines '
+                                'are required.',
+                                severity=RESULT_SEVERITY.NORMAL,
+                                file='default')],
+            filename='default')
+
+    def test_of_one_element_range(self):
```
Comment (palash25) [inline tests/general/LineCountBearTest.py]: `test_of_one_element_range`  -->  `test_one_element_range`
:+1:

File: tests/general/LineCountBearTest.py [hunk 40]
```diff
@@ -40,3 +42,52 @@ def test_exclude_blank_lines(self):
...
+        self.check_results(
+            self.uut, [''],
+            [Result.from_values('LineCountBear',
+                                'This file has 0 lines, while 1 lines '
+                                'are required.',
+                                severity=RESULT_SEVERITY.NORMAL,
+                                file='default')],
+            filename='default')
+
+    def test_max_limit_is_smaller_than_min_limit(self):
```
Comment (palash25) [inline tests/general/LineCountBearTest.py]: `def test min_limit_exceeds_max_limit` might be better.

**Comment by damian1996** [comment]
@javydb 
Why default for min_lines_per_file should be 1? In my opinion 0 is good choice, user can change this value later.

**Comment by jayvdb** [comment]
1 and 1000 are a sensible range.
files with 0 lines usually shouldnt be part of a project.  It typically indicates the file is wrong.

There is a special case for python `__init__.py` files, as they often are zero lines.
But that is a python specific thing; most other languages do not have anything similar, and many language actually barf on a zero line file.

**Comment by jayvdb** [comment]
You have a conflict, because some whitespace was added to the args.  You will need to rebase manually.

**Comment by damian1996** [comment]
@jayvdb Yes, thanks :) Can you look at travis failure in my last commit? I didn't find anything in the log. :/

File: bears/general/LineCountBear.py [hunk 41]
```diff
@@ -17,20 +18,30 @@ def _get_blank_line_count(self, file):
             list(filter(lambda x: re.match(r'^\s*$', x), file)))
         return num_blank_lines
 
-    def run(self, filename, file, max_lines_per_file: int,
+    def run(self, filename, file, min_lines_per_file: int=1,
+            max_lines_per_file: int=1000,
```
Comment (jayvdb) [inline bears/general/LineCountBear.py]: you need to amend the commits so that each default has spaces around the `=`

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala-bears/pull/2227/commits/52b516deb4bb0c63acef61f61207779fc2a22326 https://github.com/coala/coala-bears/pull/2227/commits/869e210eff72d6f524f2255b7c6e7a2d6d126d33 https://github.com/coala/coala-bears/pull/2227/commits/a48861b8c93c2f33bc5e8532affaa9f6e4e1a91e https://github.com/coala/coala-bears/pull/2227/commits/c086e4ab248c54962ba9aca8f50102e9c2205064 https://github.com/coala/coala-bears/pull/2227/commits/20cdaf4e249cfd41029aaa804c7c58b747b05146

**Comment by jayvdb** [comment]
@gitmate-bot ff
